substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs

```diff
@@ -1395,6 +1395,84 @@ where
 		}
 	}
 
+	/// Fetches the 'provides' tags associated to transactions part of blocks
+	/// of an enacted fork.
+	///
+	/// The 'provides' tags of transactions from enacted blocks are searched
+	/// in inactive views. These tags serve as cache based on the inactive views,
+	/// since prunning transactions is based on their 'provides' tags, which if
+	/// not found in the enacted fork tip active view, then they will be obtained
+	/// by revalidating the transaction, which takes time.
+	async fn provides_tags_from_inactive_views(
+		&self,
+		tree_route: &TreeRoute<Block>,
+	) -> HashMap<ExtrinsicHash<ChainApi>, Vec<Tag>> {
+		// For every enacted block, take its txs.
+		let blocks_xts = future::join_all(
+			tree_route
+				.enacted()
+				.iter()
+				.map(|hn| {
+				    let api = self.api.clone();
+					async move {
+						api
+							.block_body(hn.hash)
+							.await
+							.unwrap_or_else(|e| {
+								log::warn!(target: LOG_TARGET, "provides_tags_from_inactive_views: block_body error request: {}", e);
+								None
+							})
+							.unwrap_or_default()
+					}
+				}))
+			.await
+			.into_iter()
+			.collect::<Vec<_>>();
+
+		// For each enacted block transaction, look for its provides tags in
+		// inactive views of blocks on the retracted fork. It is possible that
+		// transactions of the enacted block to show up in multiple inactive
+		// views, but we'll consider the tags from most recent inactive view.
+		let mut provides_tags_map = HashMap::new();
+		blocks_xts.iter().for_each(|xts| {
+			tree_route
+				.retracted()
+				.iter()
+				// Skip the tip of the fork, since its view is active, and it will be checked
+				// later when we'll prune the enacted block txs, reported to the active view,
+				// which is cloned from the tip of the retracted fork.
+				.skip(1)
+				.chain(std::iter::once(tree_route.common_block()))
+				.rev()
+				.for_each(|hn| {
+					let view = self.view_store.get_view_at(hn.hash, true);
+					// Map to transaction hashes, getting an empty list if
+					// there is no view for the given block hash.
+					let xts_hashes = xts
+						.iter()
+						.filter_map(|hash| view.as_ref().map(|(inner, _)| inner.pool.hash_of(hash)))
+						.collect::<Vec<_>>();
+					// Get tx provides tags from inactive view's pool.
+					let provides_tags = view
+						.map(|(inner, _)| inner.pool.validated_pool().extrinsics_tags(&xts_hashes))
+						.unwrap_or(vec![None; xts_hashes.len()]);
```

Comment:
If there is no view for given hash we should just skip the iteration for view. Shall be simpler to follow.

Reply:
Not sure why you resolved this? We should not instantiate empty vec here.